keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1162 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Add fine-grained authorization checks to mutating kubernetes api calls.
+
+"fine-grained" means giving the ability to certain actors (users, groups,
+controllers, etc) to edit only particular fields, instead of entire objects.
+Additionally, it involves adding the ability to restrict existing board
+permissions to existing fields; such that if desired, we can introduce new
+fields and not grant broad permissions on them by default.
+
+
+## Motivation
+
+We wish to [add a new field](https://github.com/kubernetes/enhancements/pull/2840),
+but restrict the actors who can write to it by default.
+
+If this ability had existed in the past, we would have used it for `.metadata.finalizers`
+and likely for `.metadata.ownerReferences`
+
+The problem is that setting these fields has implications for other objects,
+potentially objects which a different set of users have permission on.
+
+Until now we have used subresources to solve this problem, but we have realized
+that they have a number of problems (see alternative design section) and it is
+best not to add additional uses of that pattern.
+
+### Goals
+
+* Add a mechanism for granting certain users generic ability to set
+  (write/update) certain fields, without granting them the ability to set other
+  fields in objects.
+* Add a mechanism for restricting certain fields from users otherwise having
+  general mutating permissions on an object type.
+
+### Non-Goals
+
+* Changing permissions for existing fields. This feature should have no
+  observable changes unless configured by administrators.
+
+## Proposal
+
+Let's introduce the user stories before the design:
+
+### User Stories
+
+#### Story 1
+
+I wish to restrict an operator "supersafe" to write only:
+
+* labels with keys beginning with "super.safe.com/"
+* annotations with keys beginning with "super.safe.com/"
+* finalizers with the prefix "super.safe.com/"
+* conditions with the type "SuperSafe"
+
+#### Story 2
+
+I'm adding a new field to pod.spec, "triggerSelfDestructOnSecurityBreach".
+Ordinary users must not be able to set this to true. Only admins may set it to
+true.
+
+#### List of known fields wanting these features
```

Comment:
We have also had CVEs in the past which would have been better solved by a field-level mechanism, I think, such as `Service.spec.externalIPs`